applications/serial_lte_modem/src/gnss/slm_at_gnss.c

```diff
@@ -0,0 +1,741 @@
+		}
+	} else if (run_type == RUN_TYPE_PGPS) {
+		err = nrf_cloud_pgps_process(evt->data.msg.buf, evt->data.msg.len);
+		if (err) {
+			LOG_ERR("Unable to process P-GPS data, error: %d", err);
+		}
+	} else if (run_type == RUN_TYPE_CELL_POS) {
+		struct nrf_cloud_cell_pos_result result;
+
+		err = nrf_cloud_cell_pos_process(evt->data.msg.buf, &result);
+		if (err == 0) {
+			sprintf(rsp_buf, "\r\n#XCELLPOS: %d,%lf,%lf,%d\r\n",
+				result.type, result.lat, result.lon, result.unc);
+			rsp_send(rsp_buf, strlen(rsp_buf));
+		} else if (err == 1) {
+			LOG_WRN("No position found");
+		} else {
+			LOG_ERR("Unable to process cell pos data, error: %d", err);
+		}
+	}
+
+	LOG_DBG("%s result: %d", __func__, err);
+}
+
+static void cloud_event_handler(const struct cloud_backend *const backend,
+				const struct cloud_event *const evt, void *user_data)
+{
+	ARG_UNUSED(backend);
+	ARG_UNUSED(user_data);
+
+	switch (evt->type) {
+	case CLOUD_EVT_CONNECTING:
+		LOG_DBG("CLOUD_EVT_CONNECTING");
+		break;
+	case CLOUD_EVT_CONNECTED:
+		LOG_INF("CLOUD_EVT_CONNECTED");
+		break;
+	case CLOUD_EVT_READY:
+		LOG_INF("CLOUD_EVT_READY");
+		nrf_cloud_ready = true;
+		on_cloud_evt_ready();
+		break;
+	case CLOUD_EVT_DISCONNECTED:
+		LOG_INF("CLOUD_EVT_DISCONNECTED");
+		break;
+	case CLOUD_EVT_ERROR:
+		LOG_ERR("CLOUD_EVT_ERROR");
+		break;
+	case CLOUD_EVT_DATA_SENT:
+		LOG_DBG("CLOUD_EVT_DATA_SENT");
+		break;
+	case CLOUD_EVT_DATA_RECEIVED:
+		LOG_INF("CLOUD_EVT_DATA_RECEIVED");
+		on_cloud_evt_data_received(evt);
+		break;
+	case CLOUD_EVT_PAIR_REQUEST:
+		LOG_DBG("CLOUD_EVT_PAIR_REQUEST");
+		break;
+	case CLOUD_EVT_PAIR_DONE:
+		LOG_DBG("CLOUD_EVT_PAIR_DONE");
+		break;
+	case CLOUD_EVT_FOTA_DONE:
+		LOG_DBG("CLOUD_EVT_FOTA_DONE");
+		break;
+	case CLOUD_EVT_FOTA_ERROR:
+		LOG_ERR("CLOUD_EVT_FOTA_ERROR");
+		break;
+	default:
+		break;
+	}
+}
+#endif
+
+/**@brief handle AT#XGPS commands
+ *  AT#XGPS=<op>[,<interval>[,<timeout>]]
+ *  AT#XGPS? READ command not supported
+ *  AT#XGPS=?
+ */
+int handle_at_gps(enum at_cmd_type cmd_type)
+{
+	int err = -EINVAL;
+	uint16_t op;
+	uint16_t interval;
+	uint16_t timeout;
+
+	switch (cmd_type) {
+	case AT_CMD_TYPE_SET_COMMAND:
+		err = at_params_unsigned_short_get(&at_param_list, 1, &op);
+		if (err < 0) {
+			return err;
+		}
+		if (op == GPS_START) {
+			err = at_params_unsigned_short_get(&at_param_list, 2, &interval);
+			if (err < 0) {
+				return err;
+			}
+			/* GNSS API spec check */
+			if (interval != 0 && interval != 1 &&
+			   (interval < 10 || interval > 65535)) {
+				return -EINVAL;
+			}
+			err = nrf_modem_gnss_fix_interval_set(interval);
+			if (err) {
+				LOG_ERR("Failed to set fix interval, error: %d", err);
+				return err;
+			}
+			if (at_params_unsigned_short_get(&at_param_list, 3, &timeout) == 0) {
+				err = nrf_modem_gnss_fix_retry_set(timeout);
+				if (err) {
+					LOG_ERR("Failed to set fix retry, error: %d", err);
+					return err;
+				}
+			}
+			run_type = RUN_TYPE_GPS;
```

Comment:
is there any protection from the user sending one type of command, and then, before it completes, sending another? 
for example, if i send an A-GPS command, `run_type` gets set to `RUN_TYPE_AGPS`.
then i immediately send a cell pos command, will `run_type` be set `RUN_TYPE_CELL_POS`?

Reply:
No protection for this kind of usage, but I'll add a line in the documentation

Reply:
i think it would be better to check `run_type != RUN_TYPE_NONE` when the command is issued and return failure (busy?) to the user.
you would need to ensure that `run_type` is reset to `RUN_TYPE_NONE` when a "run" is completed.

Reply:
Thanks, added basic protection in AT command handlers.